sd/endpointer.go

```diff
@@ -0,0 +1,75 @@
+package sd
+
+import (
+	"io"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	iEndpoint "github.com/go-kit/kit/sd/internal/endpoint"
+)
+
+// Endpointer listens to a service discovery system and yields a set of
+// identical endpoints on demand. An error indicates a problem with connectivity
+// to the service discovery system, or within the system itself; an Endpointer
+// may yield no endpoints without error.
+type Endpointer interface {
+	Endpoints() ([]endpoint.Endpoint, error)
```

Comment:
@peterbourgon Take a look at the latest commit. I changed `Instancer` from pushing `[]string` to  pushing `Event` struct that contains either instances or an error. `consul` implementation updated accordingly.

I have not introduced any special error, it is simply whichever error the underlying sd system returns, because it may be difficult to find common error patterns across all sd implementations. The semantics are simple - if a notification contains an error, previously pushed instances are unreliable, stale, "use at your own risk".

This leaves one open question. Typically discovery middleware should avoid breaking the application when temporarily loosing connection to discovery backend. However, in the current form loosing a connection will trigger a push of an Event with the error, and the `endpointCache` will close all Endpoints, even though the actual service instances are probably just fine at their previously known locations. Previous pull model did not have that behavior because all errors from SD backend were simply logged, without erasing the Endpoints. We can restore that (desired) behavior by not closing Endpoints in the `endpointCache` when an error is received, however that effectively puts us back to the state the code was in yesterday, i.e. the Endpointer always returns Endpoints and never returns the error. If we change the API contract to say that Endpointer may return _both_ Endpoints _and_ error, then it's a breaking change, and in particular it breaks the lb implementations since they bail on error right away.

Thus, while I like the latest commit, I am back to my suggestion of removing the `error` from the `Endpointer` signature - it doesn't do any good. If the application needs to be notified of an issue with discovery system and take some measures, then it can easily register another channel on the `Instancer`. Maybe I am missing some use cases (like a decorator Endpointer that does health checks - but that seems like it should sit between Instancer and endpointCache).